File: android/src/main/java/com/octopuscommunity/octopusreactnativesdk/CommunityDataMapper.kt
package com.octopuscommunity.octopusreactnativesdk

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.octopuscommunity.sdk.domain.model.OctopusCommunityData

/**
 * Serializes [OctopusCommunityData] to the JS-facing `OctopusCommunityData` shape.
 *
 * The native property is named `userId`; the JS type calls it `profileId` (matching the
 * cross-platform wrapper naming already used for the Flutter SDK) — this mapper is where that
 * rename happens.
 */
object CommunityDataMapper {

  fun toWritableMap(data: OctopusCommunityData): WritableMap {
    val map = Arguments.createMap()
    map.putString("profileId", data.userId)
    // Bound to a local val before the null check: `data.messageCount` is a property declared in
    // a different module (the SDK library), so Kotlin cannot smart-cast it directly even right
    // after a null check on the property access itself.
    val messageCount = data.messageCount
    if (messageCount != null) {
      map.putInt("messageCount", messageCount)
    } else {
      map.putNull("messageCount")
    }
    val gamification = data.gamification
    if (gamification != null) {
      val gamificationMap = Arguments.createMap()
      gamificationMap.putInt("level", gamification.level)
      val score = gamification.score
      if (score != null) {
        gamificationMap.putInt("score", score)
      } else {
        gamificationMap.putNull("score")
      }
      map.putMap("gamification", gamificationMap)
    } else {
      map.putNull("gamification")
    }
    return map
  }
}

File: ios/CommunityDataMapper.swift
import Octopus

/// Serializes `OctopusCommunityData` to the JS-facing `OctopusCommunityData` shape.
///
/// Unlike the Android mapper, no rename happens here: the native property is already named
/// `profileId`, matching the JS type (see `types/octopusCommunityData.ts`).
struct CommunityDataMapper {

  static func toDictionary(_ data: OctopusCommunityData) -> [String: Any] {
    var map: [String: Any] = [
      "profileId": data.profileId,
      "messageCount": (data.messageCount as Any?) ?? NSNull(),
    ]
    if let gamification = data.gamification {
      map["gamification"] = [
        "level": gamification.level,
        "score": (gamification.score as Any?) ?? NSNull(),
      ]
    } else {
      map["gamification"] = NSNull()
    }
    return map
  }
}
